core/src/main/java/jenkins/model/logging/LogStorage.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2016-2018 CloudBees Inc., Google Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model.logging;
+
+import hudson.Launcher;
+import hudson.console.AnnotatedLargeText;
+import hudson.model.BuildListener;
+import hudson.model.Node;
+import hudson.model.Run;
+import javax.annotation.CheckForNull;
+import javax.annotation.CheckReturnValue;
+import javax.annotation.Nonnull;
+
+import hudson.model.TaskListener;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.Beta;
+import org.kohsuke.stapler.export.Exported;
+import org.kohsuke.stapler.export.ExportedBean;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.util.List;
+
+/**
+ * Defines Log storage for Jenkins.
+ * The log storage may store logs for {@link Run}s and other {@link Loggable} types.
+ * This abstract class defines a low-level API which should be overridden by implementations
+ * to externalize logs.
+ * {@link LogStorageFactory} allows locating log storages for particular components.
+ *
+ * @author Oleg Nenashev
+ * @author Xing Yan
+ * @see LogStorageFactory
+ * @since TODO
+ */
+@Restricted(Beta.class)
+@ExportedBean
+public abstract class LogStorage<T extends Loggable> {
+
+    protected transient T loggable;
```

Comment:
Better to delete this and pass a `T` to any method which might need it. Make this an `interface` while you are at it.